effectful/handlers/llm/type_synthesis.py

```diff
@@ -0,0 +1,327 @@
+"""Type/class synthesis for LLM-generated code."""
+
+import ast
+import collections
+import collections.abc
+import ctypes
+import dataclasses
+import inspect
+import linecache
+import sys
+import textwrap
+import types
+import typing
+
+
+class _PyMappingProxyObject(ctypes.Structure):
+    """Internal ctypes structure to access the underlying dict of a mappingproxy."""
+
+    _fields_ = [
+        ("ob_refcnt", ctypes.c_ssize_t),
+        ("ob_type", ctypes.py_object),
+        ("mapping", ctypes.py_object),
+    ]
+
+
+import pydantic
+from litellm import OpenAIMessageContentListBlock
+from pydantic import Field
+
+from effectful.handlers.llm import LexicalContext, Template
+from effectful.handlers.llm.encoding import EncodableAs, type_to_encodable_type
+from effectful.handlers.llm.synthesis import (
+    EncodableLexicalContext,
+    SynthesisError,
+    run_mypy_check,
+)
+from effectful.ops.semantics import fwd
+from effectful.ops.syntax import ObjectInterpretation, implements
+
+
+class SynthesizedType(pydantic.BaseModel):
+    """Structured output for type/class synthesis.
+
+    Pydantic model representing synthesized class code with type name and module code.
+    """
+
+    type_name: str = Field(
+        ...,
+        description="The name of the class that satisfies the specification",
+    )
+    module_code: str = Field(
+        ...,
+        description="Complete Python module code with the class definition (no imports needed)",
+    )
+
+
+@type_to_encodable_type.register(type)
+class EncodableSynthesizedType(
+    EncodableAs[type, SynthesizedType],
+):
+    """Encodes type to SynthesizedType and vice versa."""
+
+    t = SynthesizedType
+
+    @classmethod
+    def encode(cls, vl: type, context: LexicalContext | None = None) -> SynthesizedType:
+        """Encode a type to a SynthesizedType.
+
+        Extracts the type name and source code.
+        """
+        type_name = vl.__name__
+        try:
+            source = inspect.getsource(vl)
+        except (OSError, TypeError):
+            # If we can't get source, create a minimal representation
+            source = f"class {type_name}: pass  # Source unavailable"
+
+        return SynthesizedType(
+            type_name=type_name, module_code=textwrap.dedent(source).strip()
+        )
+
+    # Counter for unique filenames
+    _decode_counter: typing.ClassVar[int] = 0
+
+    @classmethod
+    def decode(cls, vl: SynthesizedType, context: LexicalContext | None = None) -> type:
```

Comment:
There's some really scary stuff happening in here. Is the sole aim to support calling `inspect.getsource` on the result of `decode`?